File: src/test/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/ExternalMovementRepositoryTest.java
package uk.gov.justice.hmpps.prison.repository.jpa.repository;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.test.context.ActiveProfiles;
import uk.gov.justice.hmpps.prison.repository.jpa.model.MovementDirection;
import uk.gov.justice.hmpps.prison.repository.jpa.model.MovementType;

import java.time.Clock;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.AssertionsForClassTypes.tuple;
import static org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase.Replace.NONE;

@DataJpaTest
@ActiveProfiles("test")
@AutoConfigureTestDatabase(replace = NONE)
public class ExternalMovementRepositoryTest {

    @Autowired
    public ExternalMovementRepository externalMovementRepository;

    @Autowired
    public ReferenceCodeRepository<MovementType> movementTypeRepository;

    @Test
    public void queryShouldReturnExpectedResult() {
        final var temporaryAbsenceMovementType = movementTypeRepository.findById(MovementType.TAP).orElseThrow();
        Clock clock = Clock.fixed(
            LocalDateTime.of(2020, 1, 2, 3, 4, 5).atZone(ZoneId.systemDefault()).toInstant(),
            ZoneId.systemDefault());

        LocalDate currentDate = LocalDate.now(clock);

        final var absences = externalMovementRepository.findCurrentTemporaryAbsencesForPrison("LEI", temporaryAbsenceMovementType, currentDate.minusYears(1));
        assertThat(absences).hasSize(1);
        assertThat(absences).extracting(
            "offenderBooking.bookingId",
            "movementDirection",
            "movementType.code"
        ).containsExactly(tuple(-25L, MovementDirection.OUT, "TAP"));
    }
}

File: src/test/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/ExternalMovementRepositoryTest.kt
package uk.gov.justice.hmpps.prison.repository.jpa.repository

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.AssertionsForClassTypes.tuple
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.test.context.ActiveProfiles
import uk.gov.justice.hmpps.prison.repository.jpa.model.MovementDirection
import uk.gov.justice.hmpps.prison.repository.jpa.model.MovementType

@DataJpaTest
@ActiveProfiles("test")
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
class ExternalMovementRepositoryTest {
  @Autowired
  private lateinit var externalMovementRepository: ExternalMovementRepository

  @Autowired
  private lateinit var movementTypeRepository: ReferenceCodeRepository<MovementType>

  @Test
  fun queryShouldReturnExpectedResult() {
    val temporaryAbsenceMovementType = movementTypeRepository.findById(MovementType.TAP).orElseThrow()
    val absences =
      externalMovementRepository.findCurrentTemporaryAbsencesForPrison("LEI", temporaryAbsenceMovementType)
    assertThat(absences).hasSize(1)
    assertThat(absences).extracting(
      "offenderBooking.bookingId",
      "movementDirection",
      "movementType.code",
    ).containsExactly(tuple(-25L, MovementDirection.OUT, "TAP"))
  }
}
